text/0000-array-private-classes

```diff
@@ -0,0 +1,148 @@
+- Feature Name: collections should not expose internal classes through iterators
+- Start Date: 2022-01-08
+- RFC PR: (leave this empty)
+- Pony Issue: (leave this empty)
+
+# Summary
+
+Following information hiding design principle, the builtin classes of collection data structures must not be made visible through iterator functions like `ArrayKeys`, `ArrayValues` and `ArrayPairs`. These classes can be made private as they are only used as return types for `Array` functions `keys`, `values` and `pairs`. The return values for these functions are changed to the more general interface `Iterator`.
+
+A new interface `RewindableIterator` is defined to allow for rewindable iterators, like it is the case for `Array` `values`.
+
+This design principle is applied to the other collection classes that expose internals too like:
+
+* `List`
+* `Map`
+* persistent `Map`
+* `Vec`
+* persistent `Vec`
+* `Set`
+* `Itertools`
+
+# Motivation
+
+This change brings:
+
+- Applying the design principle of [hiding implementation details](https://en.wikipedia.org/wiki/Information_hiding) but offer a general and stable interface. Returning interfaces instead of concrete classes allows changing the implementation. Usually, one must return the most general type that fullfils the contract of the function (in the case of the functions discussed in this RFC, iteration).
+- Collections' functions `keys`, `values` and `pairs` definitions are made more general. Iterators implementation details are not public. Internal classes used by implementation like `*Keys`, `*Values` and `*Pairs` are now [opaque data types](https://en.wikipedia.org/wiki/Opaque_data_type). Generally, when using these collection classes, clients are not interested by the iterators implementation, but by the types these iterators return and that is provided by the generic parameters.
+- The generic return signature of these 3 iterating functions is simpler to understand for clients of collection classes.
+- Makes the standard library more simple by hiding 18 specialised classes from stdlib of which 3 are from `builtin`.
+- The interface `RewindableIterator` is added to create rewindable iterators (can be re-start from first value).
+
+This change remains compatible with the existing code base but for client code that is directly using the classes `*Keys`, `*Values` and `*Pairs`. A search on Github shows that the impact is very limited.
+
+# Detailed design
```

Comment:
I'd like to see a full enumeration of all the existing method signatures that would change and what they would change to.

Reply:
You can see them on [my branch](https://github.com/pmetras/ponyc/tree/rfc-array).

Reply:
Please include them in the RFC.

Reply:
Done